tests/test_assembly.py

```diff
@@ -464,5 +465,67 @@ def threaded_nonsym(u, v, w):
         )
 
 
+@pytest.mark.parametrize(
+    "m,mdgtype,etype",
+    [
+        (
+            MeshTri.init_tensor(np.linspace(0, 1, 7),
+                                np.linspace(0, 1, 7)),
+            MeshTri1DG,
+            ElementTriP1,
+        ),
+        (
+            MeshTri.init_tensor(np.linspace(0, 1, 5),
+                                np.linspace(0, 1, 5)),
+            MeshTri1DG,
+            ElementTriP1,
+        ),
+        (
+            MeshTri().refined(2),
+            MeshTri1DG,
+            ElementTriP1,
+        ),
+        (
+            MeshTri().refined(3),
+            MeshTri1DG,
+            ElementTriP1,
+        ),
+        (
+            MeshQuad().refined(2),
+            MeshQuad1DG,
+            ElementQuad1,
+        ),
+        (
+            MeshQuad().refined(2),
+            MeshQuad1DG,
+            ElementQuad2,
+        ),
+        (
+            MeshTri().refined(2),
```

Comment:
There is a caveat that no element can touch both of the periodic interfaces. I must figure out a way to detect that. Maybe by checking for duplicate node indices over each element?

Reply:
It does seem unlikely in practice, having a single element spanning the entire domain in one direction, but it is good practice to cover these degenerate cases where possible.

I notice that `test_periodic_mesh_assembly` doesn't test `MeshLine1DG`, but then that would be because it hasn't been written yet.  Actually that's the one I wanted first.  I'll try adding that as an exercise.

Reply:
Well, that particular test won't work in one dimension because the Poisson&ndash;Dirichlet problem is undefined, but anyway.  A simple test might be the eigenvalues of the laplacian on a line-segment, with the modes being the sines and cosines?

More generally, I'm a little uncomfortable with `test_periodic_mesh_assembly` as if the implementation of periodicity didn't actually work at all but just left the nominally periodic boundaries as default homogeneous Neumann, these tests would pass&mdash;the solutions are all going to be independent of the periodic coordinate, aren't they?  Because the forcing is

https://github.com/kinnala/scikit-fem/blob/4a295b519510ff7db7ab7df6aa4ab875116b4406/tests/test_assembly.py#L557

I'm not saying that I doubt the implementation, just that the test doesn't test it and if the implementation gets broken later, we might not be notified.

Also do we want an example as well as a suite of tests?  I might try that, if I get `MeshLine1DG` up.

Besides the advection&ndash;diffusion example that I posted earlier, which isn't a great test, I wonder maybe what about a periodic forcing?  That might be less trivial: I mean the derivatives at the periodic boundary should be nonzero.

Say u (x) = sin (2 &pi; _x_), u' = 2 &pi; cos 2 &pi; _x_, u'' = -4 &pi;&sup2; sin 2 &pi; _x_, so force with sin 2&pi;_x_ on 0 < *x* < 1?

Reply:
No, I'm afraid I can't get `MeshLine1DG` going; I find that I don't understand the implementation.

I assumed
```python
class ElementLineP1DG(ElementLineP1):

    nodal_dofs = 0
    interior_dofs = 2
    dofnames = ['u'] * 2
...
class MeshLine1DG(MeshLine1):
    """A lineal mesh with a discontinuous topology."""
    elem: Type[Element] = ElementLineP1DG
    affine: bool = False
    sort_t: bool = False
```
but then
```python
m = MeshLine([0, 1, 2])
mp = MeshLine1DG.periodic(m, [-1], [0])
b = Basis(mp, ElementLineP1())
```
fails with 
```
/home/gdmcbain/src/scikit-fem/skfem/assembly/basis/abstract_basis.py:62: UserWarning: Unable to calculate DOF locations.
  warnings.warn("Unable to calculate DOF locations.")
Traceback (most recent call last):
  File "/home/gdmcbain/src/scikit-fem/mwe.py", line 9, in <module>
    b = Basis(mp, elem)
  File "/home/gdmcbain/src/scikit-fem/skfem/assembly/basis/cell_basis.py", line 68, in __init__
    self.basis = [self.elem.gbasis(self.mapping, self.X, j, tind=elements)
  File "/home/gdmcbain/src/scikit-fem/skfem/assembly/basis/cell_basis.py", line 68, in <listcomp>
    self.basis = [self.elem.gbasis(self.mapping, self.X, j, tind=elements)
  File "/home/gdmcbain/src/scikit-fem/skfem/element/element_h1.py", line 13, in gbasis
    invDF = mapping.invDF(X, tind)
  File "/home/gdmcbain/src/scikit-fem/skfem/mapping/mapping_isoparametric.py", line 175, in invDF
    J = [[self.J(i, j, X, tind=tind) for j in range(self.dim)]
  File "/home/gdmcbain/src/scikit-fem/skfem/mapping/mapping_isoparametric.py", line 175, in <listcomp>
    J = [[self.J(i, j, X, tind=tind) for j in range(self.dim)]
  File "/home/gdmcbain/src/scikit-fem/skfem/mapping/mapping_isoparametric.py", line 175, in <listcomp>
    J = [[self.J(i, j, X, tind=tind) for j in range(self.dim)]
  File "/home/gdmcbain/src/scikit-fem/skfem/mapping/mapping_isoparametric.py", line 70, in J
    out += p[i, t[itr, :]][:, None] * dphi[j]
IndexError: index 4 is out of bounds for axis 1 with size 4
```

Indeed, `mp.doflocs.shape` is `(1, 4)` but `mp.dofs.element_dofs` is
```python
array([[0, 2, 4],
       [1, 3, 5]])
```
so that the indices of the last column are `> mp.doflocs.shape[1]`.  Why does `mp.t` have three elements instead of two like `m.p`?  When I use `MeshTri1DG`, the shape of `.t` doesn't change from `m` to `mp`.

Reply:
Try to refine `MeshLine` once. Now you have two equal elements and I think some of the indexing routines cannot deal with that.

Reply:
Somethings not right, I need to look into this.

Reply:
> Try to refine MeshLine once. Now you have two equal elements

How do you mean?  `MeshLine().refined()` has p `[0, 1, 0.5]` and t `[[0, 2], [2, 1]]`, which seems correct?  The two elements are equal in length, but that doesn't seem like it should cause a problem.  What do you mean by &lsquo;equal&rsquo;?

Reply:
This was in your example: `m = MeshLine([0, 1, 2])`. If you periodify that, then you have elements `[0 1]` and `[1 0]` which does not work for the same reason that this thread was started because there is no check yet. In theory it should work but I think some of the indexing routines will consider those elements equal.

Reply:
Oh, you found a bug which was caused by having [0] in one of the sets.

Reply:
Another bug lurking somewhere.

Reply:
`MeshLine1.__post_init__` looks suspicious.

Reply:
Alright, that if-clause won't work here.

Reply:
This now looks better:
```python
from skfem import *
from typing import Type

class ElementLineP1DG(ElementLineP1):

    nodal_dofs = 0
    interior_dofs = 2
    dofnames = ['u'] * 2

class MeshLine1DG(MeshLine):
    """A lineal mesh with a discontinuous topology."""
    elem: Type[Element] = ElementLineP1DG
    affine: bool = False
    sort_t: bool = False

    def __post_init__(self):
        pass

m = MeshLine().refined(2)
mp = MeshLine1DG.periodic(m, [0], [1])
```

Reply:
> `MeshLine1.__post_init__` looks suspicious.

Can we move this initialization with 1D list to another constructor? `MeshLine1.init_linear([0, 1, 2])` or something like that.

Reply:
I have another idea: make `MeshLine` not a shorthand for `MeshLine1` but a function which will support these fancy alternative constructor parameters.

Reply:
That's a good idea.  Yes, this is #663 again.  I so often turn to the one-dimensional case thinking that it will be easier and find special obstacles.

Reply:
The next hurdle is that `MappingIsoparametric` isn't implemented in one dimension.

I copied `affine = False` from `MeshTriP1DG` without knowing its rationale.

```
Traceback (most recent call last):
  File "/home/gdmcbain/src/scikit-fem/mwe.py", line 22, in <module>
    b = Basis(mp, elem)
  File "/home/gdmcbain/src/scikit-fem/skfem/assembly/basis/cell_basis.py", line 68, in __init__
    self.basis = [self.elem.gbasis(self.mapping, self.X, j, tind=elements)
  File "/home/gdmcbain/src/scikit-fem/skfem/assembly/basis/cell_basis.py", line 68, in <listcomp>
    self.basis = [self.elem.gbasis(self.mapping, self.X, j, tind=elements)
  File "/home/gdmcbain/src/scikit-fem/skfem/element/element_h1.py", line 13, in gbasis
    invDF = mapping.invDF(X, tind)
  File "/home/gdmcbain/src/scikit-fem/skfem/mapping/mapping_isoparametric.py", line 177, in invDF
    detDF = self.detDF(X, tind, J=J)
  File "/home/gdmcbain/src/scikit-fem/skfem/mapping/mapping_isoparametric.py", line 165, in detDF
    raise Exception("Not implemented for the given dimension.")
Exception: Not implemented for the given dimension.
```

Reply:
I hit the lack of rhe isoparamereic line rhe other day in discussing the speed benchmark #670 and the erroneous laplacian for the affine rectangular quad.  I had thought that isooarametric would reduce to affine for a rectangular quad and tried to investigate rhat by looking at the assumed simpler but in fact unfortunately anomalous line.

Reply:
> I have another idea: make MeshLine not a shorthand for MeshLine1 but a function which will support these fancy alternative constructor parameters.

The more i think about this, rhe better an idea it seems. The use of `dataclasses` is fouled by coercive constructors, but that all goes away if we use a constructor function. Then the instances can always be instantiated with their attribures without having to modify them. The `MeshLine` doesn't have to be a class, it can just be a function, like `numpy.array` which constructs an `ndarray`.

Reply:
> I copied `affine = False` from `MeshTriP1DG` without knowing its rationale.

That's correct. The thinking is that if something else than 1 node per vertex is used in the geometry representation, then `MappingIsoparametric` should be used. I'll try fixing `MappingIsoparametric` for 1D.